wave_kinematics/tests/rotation_test.cpp

```diff
@@ -0,0 +1,185 @@
+  // First make sure the rotation is not already identity.
+  ASSERT_FALSE(rotation1.isNear(eye3));
+  rotation1.setToIdentity();
+  ASSERT_TRUE(rotation1.isNear(eye3));
+}
+
+// Test that rotation composition returns expected product.
+
+TEST_F(RotationTestFixture, test_compose) {
+  rotation1.compose(rotation2);
+  ASSERT_TRUE(rotation1.isNear(rotation3));
+}
+
+// Test that rotation inverse returns correct matrix.
+
+TEST_F(RotationTestFixture, test_inverse) {
+  rotation1.invert();
+  ASSERT_TRUE(rotation1.isNear(rotation1_inverse));
+}
+
+// Test that rotation of a vector returns correct value.
+TEST_F(RotationTestFixture, test_rotate) {
+  wave::Vec3 v1_rotated_test = rotation1.rotate(v1);
+  wave::Vec3 error = v1_rotated_test - v1_rotated;
+  ASSERT_LE(error.norm(), wave::Rotation::comparision_threshold);
+}
+
+// Test that inverse rotation of a vector returns correct value.
+TEST_F(RotationTestFixture, test_inverse_rotate) {
+  wave::Vec3 v1_inverse_rotated_test = rotation1_inverse.inverseRotate(v1);
+  wave::Vec3 error = v1_inverse_rotated_test - v1_rotated;
+  ASSERT_LE(error.norm(), wave::Rotation::comparision_threshold);
+}
+
+// Test functionality of the boxplus operator.
+TEST_F(RotationTestFixture, test_manifold_plus) {
+  // We know that rotation3 = rotation1*rotation2
+  // Therefore rotation3 = exp(rotation1_angleaxis_params)*rotation2,
+  // which is exactly the boxplus operation.
+
+  // Make sure rotation2 and rotation3 are not the same initially
+  ASSERT_FALSE(rotation2.isNear(rotation3));
+  // Use manifold plus to make rotation2==rotation3.
+  rotation2.manifoldPlus(rotation1_angleaxis_params);
+  ASSERT_TRUE(rotation2.isNear(rotation3));
+}
+
+/* JACOBIAN TESTING */
+
+// Compare the rotateAndJacobian derivatives numerically.
+
+TEST_F(RotationTestFixture, test_rotate_and_jacobian) {
+  // First get the analytical Jacobians.
+  wave::Mat3 Jparam_analytical, Jpoint_analytical;
+  rotation1.rotateAndJacobian(v1, Jpoint_analytical, Jparam_analytical);
+
+  // Now get the numerical Jacobians
+  wave::Mat3 Jparam_numerical(3, 3), Jpoint_numerical(3, 3);
+
+  // Jacobian wrt the point.
+  wave::RotateAndJacobianJpointFunctor Jpoint_functor(rotation1);
+  wave::numerical_jacobian(Jpoint_functor, v1, Jpoint_numerical);
+  Eigen::MatrixXd error_matrix = Jpoint_analytical - Jpoint_numerical;
+  ASSERT_LE(error_matrix.norm(), wave::Rotation::comparision_threshold);
+
+  // Jacobian wrt the rotation params.
+  wave::RotateAndJacobianJparamFunctor Jparam_functor(rotation1, v1);
+  // Perturbations are about zero vector in the tangent space.
+  wave::Vec3 perturbation_vec(0, 0, 0);
+  wave::numerical_jacobian(
+    Jparam_functor, perturbation_vec, Jparam_numerical);
+  error_matrix = Jparam_analytical - Jparam_numerical;
+  ASSERT_LE(error_matrix.norm(), wave::Rotation::comparision_threshold);
+}
+
+int main(int argc, char *argv[]) {
+  testing::InitGoogleTest(&argc, argv);
+  return RUN_ALL_TESTS();
+}
```

Comment:
At current I have a general `wave_utils_tests` and `wave_optimization_tests` test runner for each module of libwave, maybe you should do the same? For example this is how I link up all my test cases into a single test runner in the `utils` module:

```
# UNIT TESTS
SET(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/tests)
FILE(COPY tests/data DESTINATION ${PROJECT_BINARY_DIR}/tests)

ADD_EXECUTABLE(
    # TARGET
    wave_utils_tests
    # TESTS
    tests/utils/config_test.cpp
    tests/utils/data_test.cpp
    tests/utils/filesystem_test.cpp
    tests/utils/math_test.cpp
    tests/utils/time_test.cpp
    # TEST RUNNER
    tests/test_runner.cpp
)
TARGET_LINK_LIBRARIES(wave_utils_tests ${TEST_DEPS})
```

Open to suggestions if you have alternatives.